tensorflow-opgen/src/TensorFlow/OpGen/ParsedOp.hs

```diff
@@ -0,0 +1,301 @@
+                deriving Eq
+
+data AttrBaseType = AttrBytes | AttrInt64 | AttrFloat | AttrBool
+                | AttrType | AttrShape | AttrTensor
+                deriving Eq
+
+-- | An input or output argument (Tensor) for an op.
+data ParsedArg = ParsedArg
+    { parsedArgName :: Name
+    , parsedArgDescription :: Text
+    , parsedArgCase :: ParsedArgCase
+    , parsedArgKind :: ArgKind
+    }
+
+data ParsedArgCase
+    = SimpleArg { argType :: ArgType }
+    | ListArg
+        { argLength :: Name  -- ^ The attribute that specifies this list's length.
+        , argType :: ArgType
+        }
+    | MixedListArg { argTypeAttr :: Name }
+        -- ^ A heterogeneous list.
+        -- TODO(judahjacobson): Implement this.
+
+-- | The type of an argument.
+data ArgType
+    = ArgTypeFixed DataType -- ^ A fixed type.
+    | ArgTypeAttr Name  -- ^ A type that depends on an attribute.
+
+data ArgKind
+    = ArgTensorRef -- Tensor Ref a
+    | ArgTensorValue -- Tensor Value a
+    | ArgTensorEither Text -- Tensor v a
+    | ArgResource -- Resource a
+
+
+makeName :: Text -> Name
+makeName n = Name
+    { haskellName = HaskellName $ fixReservedName $ lowCase n
+    , tfName = TFName n
+    }
+
+-- | Change a name so it doesn't conflict with any Haskell keywords.
+fixReservedName :: Text -> Text
+fixReservedName n
+    | n `Set.member` reservedKeywords = n <> "'"
+    | otherwise = n
+
+reservedKeywords :: Set.Set Text
+reservedKeywords = Set.fromList $
+    -- Haskell2010 keywords:
+    -- https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-180002.4
+    -- We don't include keywords that are allowed to be variable names,
+    -- in particular: "as", "forall", and "hiding".
+    [ "case"
+    , "class"
+    , "data"
+    , "default"
+    , "deriving"
+    , "do"
+    , "else"
+    , "foreign"
+    , "if"
+    , "import"
+    , "in"
+    , "infix"
+    , "infixl"
+    , "infixr"
+    , "instance"
+    , "let"
+    , "module"
+    , "newtype"
+    , "of"
+    , "then"
+    , "type"
+    , "where"
+    ]
+    ++  -- Nonstandard extensions
+    [ "mdo"   -- RecursiveDo
+    , "rec"   -- Arrows, RecursiveDo
+    , "proc"  -- Arrows
+    ]
+
+-- | Lower-case the given text.
+lowCase :: Text -> Text
+lowCase = forceCase toLower
+
+forceCase :: (Char -> Char) -> Text -> Text
+forceCase convert s = maybe "" (\(c, cs) -> Text.cons (convert c) cs)
+                      (Text.uncons s)
+
+camelCase :: Text -> Text
+camelCase s = Text.concat $ map upCase
+                          $ filter (/= "ops")
+                          $ Text.splitOn "_" s
+
+-- | Upper-case the given text.
+upCase :: Text -> Text
+upCase = forceCase toUpper
+
+
+parseOp :: OpDef -> ParsedOp
+parseOp o = ParsedOp
+    { parsedOpName = makeName $ o ^. name
+    , parsedOpSummary = o ^. summary
+    , parsedOpDescription = o ^. description
+    , parsedInputs = inputs
+    , parsedOutputs = outputs
+    , explicitInputAttrs = explicitInputs
```

Comment:
Done.  I have mixed feelings in general, but it does seem like a clear win in this case.